### PR TITLE
[DAG] Fold (and X, (bswap/bitreverse (not Y))) -> (and X, (not (bswap/bitreverse Y)))

### LINKED ISSUE #112425: Allow `AND-NOT` to be emitted from `NOT` + <some bit/byte reordering> + `AND`
Zen 5 Assembly:

```diff
foo:
        mov     rax, rdi
-       not     rax
        bswap   rax
-       and     rax, rdi
+       andn    rax, rax, rdi
        ret
```

Zig version: ([Godbolt](https://zig.godbolt.org/z/ssj5e1WzP))

```zig
export fn foo(a: u64) u64 {
    return a & ~@byteSwap(a);
}

export fn bar(a: u64) u64 {
    return andn(a, @byteSwap(a));
}
```

LLVM version (optimized): ([Godbolt](https://llvm.godbolt.org/z/1qeMzjWzP))

```llvm
define dso_local i64 @foo(i64 %0) local_unnamed_addr {
Entry:
  %1 = xor i64 %0, -1
  %2 = tail call i64 @llvm.bswap.i64(i64 %1)
  %3 = and i64 %2, %0
  ret i64 %3
}

declare i64 @llvm.bswap.i64(i64) #1

define dso_local i64 @bar(i64 %0) local_unnamed_addr {
Entry:
  %1 = tail call i64 @llvm.bswap.i64(i64 %0)
  %2 = tail call i64 asm "andn ${1}, ${2}, ${0}", "=r,r,r,~{dirflag},~{fpsr},~{flags}"(i64 %0, i64 %1) #3
  ret i64 %2
}
```

Problem occurs in `foo` even when you manually change it to:

```llvm
define dso_local i64 @foo(i64 %0) local_unnamed_addr {
Entry:
  %1 = tail call i64 @llvm.bswap.i64(i64 %0)
  %2 = xor i64 %1, -1
  %3 = and i64 %2, %0
  ret i64 %3
}
```

### PR DESCRIPTION
**Pull request by RKSimon**
On ANDNOT capable targets we can always do this profitably, without ANDNOT we only attempt this if we don't introduce an additional NOT

Fixes #112425

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -7350,6 +7350,21 @@ SDValue DAGCombiner::visitAND(SDNode *N) {
...
+  // that might have become separated by a bitwise-agnostic instruction.
+  if (TLI.hasAndNot(SDValue(N, 0))) {
+    SDValue X, Y;
+
+    // Fold (and X, (bswap (not Y))) -> (and X, (not (bswap Y)))
+    // Fold (and X, (bitreverse (not Y))) -> (and X, (not (bitreverse Y)))
+    for (unsigned Opc : {ISD::BSWAP, ISD::BITREVERSE})
+      if (sd_match(N, m_And(m_Value(X),
+                            m_OneUse(m_UnaryOp(Opc, m_Not(m_Value(Y)))))) &&
+          !sd_match(X, m_Not(m_Value())))
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7363]: It is `~X & bitreverse/bswap(~Y)`.

[the remark this answers is not in the document]
Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7363]: Yes, for the NOT(Y) we can allow multiple uses as with a ANDNOT we'll be getting it for free - I'd be willing to allow this fold to occur on targets without ANDNOT, if for those targets we require the NOT(Y) to have one use?

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7363]: Yeah, something like
```
if (sd_match(N, m_And(m_Value(X),
                            m_OneUse(m_UnaryOp(Opc, m_Value(NotY)))) &&
   match(NotY, m_Not(m_Value(Y))) && (TLI.hasAndNot(SDValue(N, 0) || NotY->hasOneUse())
```

should work.

**Comment by RKSimon** [comment]
Added non-ANDNOT handling for single use cases of the NOT

File: llvm/test/CodeGen/X86/andnot-patterns.ll [hunk 2]
```diff
@@ -529,8 +815,7 @@ define i32 @andnot_bitreverse_i32(i32 %a0, i32 %a1) nounwind {
 define i16 @andnot_bitreverse_i16(i16 %a0, i16 %a1) nounwind {
 ; X86-LABEL: andnot_bitreverse_i16:
 ; X86:       # %bb.0:
-; X86-NEXT:    movl {{[0-9]+}}(%esp), %eax
-; X86-NEXT:    notl %eax
+; X86-NEXT:    movzwl {{[0-9]+}}(%esp), %eax
```
Comment (goldsteinn) [inline llvm/test/CodeGen/X86/andnot-patterns.ll:818]: Is this equivilent?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -7350,6 +7350,17 @@ SDValue DAGCombiner::visitAND(SDNode *N) {
...
     return R;
 
+  // Fold (and X, (bswap (not Y))) -> (and X, (not (bswap Y)))
+  // Fold (and X, (bitreverse (not Y))) -> (and X, (not (bitreverse Y)))
+  SDValue X, Y, NotY;
+  for (unsigned Opc : {ISD::BSWAP, ISD::BITREVERSE})
+    if (sd_match(N,
+                 m_And(m_Value(X), m_OneUse(m_UnaryOp(Opc, m_Value(NotY))))) &&
+        sd_match(NotY, m_Not(m_Value(Y))) &&
+        (TLI.hasAndNot(SDValue(N, 0)) || NotY->hasOneUse()))
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7360]: Maybe should be `|| (NotY->hasOneUse() && sd_match(X, m_Not(m_Value()))`

There is a lot of diff from just re-assosiating the not which seems netural.

No strong opinion, however.

File: llvm/test/CodeGen/X86/andnot-patterns.ll (hunk 2 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/andnot-patterns.ll:818]: yes - we only care about this lower 16-bits but arg passing allows us to dereference 32-bits

**Comment by goldsteinn** [comment]
LGTM

**Review by dtcxzyw — APPROVED** [review]
LG

**Pull request merged**